### ISSUE TITLE
AV2220 behavior on query with multiple sources

### ISSUE THREAD

**Issue by jnyrup** [bkoelman/CSharpGuidelinesAnalyzer#137]
#### Analyzer

AvoidQuerySyntaxForSimpleExpressionAnalyzer

This is mostly a question of whether the observed behavior is by design or a bug.
In any case I'll workaround it, so don't drop everything to answer or fix/change this.

#### Repro steps

When doing a Cartesian product at least I find the query syntax less convoluted than the corresponding method syntax, but this triggers AV2220.

```cs
IEnumerable<int[]> QuerySyntax(int[] values)
{
    return
        from l in values
        from r in values
        select new[] { l, r };
}
```

```cs
IEnumerable<int[]> MethodSyntax(int[] values)
{
    return values.SelectMany(_ => values, (l, r) => new[] { l, r });
}
```

I found a unit test with multiple sources.
From the test name I would assume it was the use of multiple sources that made it "complex", but running the test it crosses the complexity threshold of 1 due to the `where`.

Removing the `where` brings the complexity score down to `1` which triggers the diagnostic.

https://github.com/bkoelman/CSharpGuidelinesAnalyzer/blob/de03d00f6120a52d35a40a29ec479a74287f623f/src/CSharpGuidelinesAnalyzer/CSharpGuidelinesAnalyzer.Test/Specs/Framework/AvoidQuerySyntaxForSimpleExpressionSpecs.cs#L388-L408

#### Expected behavior
Whenever multiple sources are used, AV2220 is not triggered.

#### Actual behavior
AV2220 is triggered.

**Comment by bkoelman** [comment]
Hi @jnyrup, thanks for reporting. This looks like a bug, indeed.

> From the test name I would assume it was the use of multiple sources that made it "complex", but running the test it crosses the complexity threshold of 1 due to the where.

You're right, but it turns out that the second `from` is a nested node, which goes uncounted. So to properly detect both `from`s, the analyzer needs to walk the tree instead of just counting the top-level clauses.

But walking the tree makes other tests fail... This would be a nice opportunity to re-rationalize how this analyzer is supposed to work. [AV2220](https://csharpcodingguidelines.com//framework-guidelines/#AV2220) isn't very well defined regarding which expressions are "simple". So when I wrote this analyzer, I made some judgment calls and tried to pin the current behavior in tests. I'd like to take to revisit those, feedback is definitely welcome.

**Comment by bkoelman** [comment]
Hi @jnyrup, would you mind [trying out](https://github.com/bkoelman/CSharpGuidelinesAnalyzer#trying-out-the-latest-build) the updated analyzer, before I release a new version?

**Comment by jnyrup** [comment]
I don't see the new artifact when connecting to the feed from "Trying out the latest build", but the code looks _much_ simpler/robust/:rocket: and `When_query_contains_multiple_sources_it_must_be_skipped` is exactly my case.

**Comment by bkoelman** [comment]
It's actually there, but under a different name (to workaround a NuGet limitation).

![image](https://user-images.githubusercontent.com/10324372/224306285-3e76e08f-aa3f-4eef-b473-987d1e75fd2f.png)

I've run it against https://github.com/fluentassertions/fluentassertions. It seems to work fine and does not introduce new warnings.

**Comment by jnyrup** [comment]
> It's actually there, but under a different name (to workaround a NuGet limitation).

ah... didn't read the guidelines _that_ closely.
